### PR TITLE
Tm 23 submit pledge

### PR DESCRIPTION
**Pull request by ConorTill**
Web/UI work for submitting a pledge.

Also see:

* https://github.com/SkillsFundingAgency/das-levy-transfer-matching-api/pull/12
* https://github.com/SkillsFundingAgency/das-apim-endpoints/pull/401

### CODE REVIEW

File: src/SFA.DAS.LevyTransferMatching.Infrastructure/Dto/AccountDto.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace SFA.DAS.LevyTransferMatching.Infrastructure.Dto
+{
+    public class AccountDto
+    {
+        public double RemainingTransferAllowance { get; set; }
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Infrastructure/Dto/AccountDto.cs:9]: Is this a monetary value, should it be decimal rather than double? Do we do monetary calculations on it or is it just an indicative value?

File: src/SFA.DAS.LevyTransferMatching.Infrastructure/Services/PledgesService/IPledgesService.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using SFA.DAS.LevyTransferMatching.Infrastructure.Dto;
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.LevyTransferMatching.Infrastructure.Services.PledgesService
+{
+    public interface IPledgesService
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Infrastructure/Services/PledgesService/IPledgesService.cs:9]: Would this be better named in the singular `IPledgeService`?

File: src/SFA.DAS.LevyTransferMatching.Infrastructure/Services/PledgesService/PledgesService.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+using SFA.DAS.LevyTransferMatching.Infrastructure.Dto;
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.LevyTransferMatching.Infrastructure.Services.PledgesService
+{
+    public class PledgesService : IPledgesService
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Infrastructure/Services/PledgesService/PledgesService.cs:12]: Would this be better named in the singular `PledgeService`?

File: src/SFA.DAS.LevyTransferMatching.Web/Helpers/UnorderedListTagHelper.cs [hunk 4]
```diff
@@ -1,26 +1,31 @@
...
 
         [HtmlAttributeName("item-class")]
         public string ItemClass { get; set; }
 
-        [ViewContext]
-        [HtmlAttributeNotBound]
-        public ViewContext ViewContext { get; set; }
+        //[ViewContext]
+        //[HtmlAttributeNotBound]
+        //public ViewContext ViewContext { get; set; }
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Helpers/UnorderedListTagHelper.cs:28]: If this isn't needed can we delete it rather than commenting out?

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 5]
```diff
@@ -28,6 +39,9 @@ public IndexViewModel GetIndexViewModel(string encodedAccountId)
         public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
         {
             var cacheItem = await RetrievePledgeCacheItem(request.CacheKey);
+            var levels = await _tagService.GetLevels();
+            var sectors = await _tagService.GetSectors();
+            var jobRoles = await _tagService.GetJobRoles();
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:44]: As there are no interdependencies these awaits could be asych rather than await'ing each individually, such as:
```suggestion
            var levelsTask = _tagService.GetLevels();
            var sectorsTask = _tagService.GetSectors();
            var jobRolesTask = _tagService.GetJobRoles();
			await Task.WhenAll(levelsTask, sectorsTask, jobRolesTask);
```

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 6]
```diff
@@ -38,46 +52,77 @@ public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
                 Sectors = cacheItem.Sectors,
                 JobRoles = cacheItem.JobRoles,
                 Levels = cacheItem.Levels,
+                LevelOptions = levels,
+                SectorOptions = sectors,
+                JobRoleOptions = jobRoles
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:57]: Linked to above comment this would now become:
```suggestion
                LevelOptions = levels.Result,
                SectorOptions = sectors.Result,
                JobRoleOptions = jobRoles.Result
```

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 7]
```diff
@@ -38,46 +52,77 @@ public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
...
                 Amount = cacheItem.Amount.ToString(),
+                RemainingTransferAllowance = remainingTransferAllowance.ToString("N0"),
                 IsNamePublic = cacheItem.IsNamePublic
             };
         }
 
         public async Task<SectorViewModel> GetSectorViewModel(SectorRequest request)
         {
             var cacheItem = await RetrievePledgeCacheItem(request.CacheKey);
+            var sectors = await _tagService.GetSectors();
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:79]: Similar to a previous comment, as these two awaits are not related they could be awaited together using WhenAll
```suggestion
            var cacheItemTask = RetrievePledgeCacheItem(request.CacheKey);
            var sectorsTask = _tagService.GetSectors();
            await Task.WhenAll(cacheItemTask, sectorsTask);
```

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 8]
```diff
@@ -38,46 +52,77 @@ public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
...
 
             return new SectorViewModel
             {
                 EncodedAccountId = request.EncodedAccountId,
                 CacheKey = request.CacheKey,
-                Sectors = cacheItem.Sectors
+                Sectors = cacheItem.Sectors,
+                SectorOptions = sectors
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:86]: Linked to previous comment:
```suggestion
                Sectors = cacheItem.Result.Sectors,
                SectorOptions = sectors.Result
```

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 9]
```diff
@@ -38,46 +52,77 @@ public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
...
             };
         }
 
         public async Task<JobRoleViewModel> GetJobRoleViewModel(JobRoleRequest request)
         {
             var cacheItem = await RetrievePledgeCacheItem(request.CacheKey);
+            var jobRoles = await _tagService.GetJobRoles();
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:93]: See previous comments on using AwaitAll

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 10]
```diff
@@ -38,46 +52,77 @@ public async Task<CreateViewModel> GetCreateViewModel(CreateRequest request)
...
+
+            if (cacheItem == null)
+            {
+                throw new InvalidOperationException("Unable to submit pledge due to cache expiry");
+            }
+
+            var pledgeDto = new PledgeDto
+            {
+                Amount = cacheItem.Amount.Value,
+                IsNamePublic = cacheItem.IsNamePublic.Value,
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:116]: Should these have a null check on them?

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs [hunk 11]
```diff
@@ -109,24 +163,17 @@ public async Task UpdateCacheItem(LevelPostRequest request)
         public async Task<LevelViewModel> GetLevelViewModel(LevelRequest request)
         {
             var cacheItem = await RetrievePledgeCacheItem(request.CacheKey);
+            var levels = await _tagService.GetLevels();
```
Comment (DannyGB) [inline src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:166]: See previous comments about using WhenAll

File: src/SFA.DAS.LevyTransferMatching.Infrastructure/Dto/AccountDto.cs (hunk 1 above)
Comment (chrisfoster76) [reply src/SFA.DAS.LevyTransferMatching.Infrastructure/Dto/AccountDto.cs:9]: I think it's defined as a double in the source (MA API)

File: src/SFA.DAS.LevyTransferMatching.Web/Helpers/UnorderedListTagHelper.cs (hunk 4 above)
Comment (chrisfoster76) [reply src/SFA.DAS.LevyTransferMatching.Web/Helpers/UnorderedListTagHelper.cs:28]: Yes, my bad

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs (hunk 5 above)
Comment (chrisfoster76) [reply src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:44]: Yes we should

File: src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs (hunk 10 above)
Comment (chrisfoster76) [reply src/SFA.DAS.LevyTransferMatching.Web/Orchestrators/PledgeOrchestrator.cs:116]: Not sure they can be null due to validation, but it might be worth trapping and throwing our own exception rather than getting a NRE

**Review by DannyGB — APPROVED** [review]
